File: filacircular.c
#include <stdio.h>

#define MAX 5

int dados[MAX];
int head     = 0;   /* índice de onde sai */
int tail     = 0;   /* índice de onde entra */
int tamanho  = 0;

/* ----- funções ----- */

int vazia() {
    return tamanho == 0;
}

int cheia() {
    return tamanho == MAX;
}

void enqueue(int valor) {
    if (cheia()) { printf("Fila cheia!\n"); return; }
    dados[tail] = valor;
    tail = (tail + 1) % MAX;   /* avança em círculo */
    tamanho++;
}

int dequeue() {
    if (vazia()) { printf("Fila vazia!\n"); return -1; }
    int valor = dados[head];
    head = (head + 1) % MAX;   /* avança em círculo */
    tamanho--;
    return valor;
}

int frente() {
    if (vazia()) { printf("Fila vazia!\n"); return -1; }
    return dados[head];
}

void imprimir() {
    printf("Fila circular (%d/%d): ", tamanho, MAX);
    for (int i = 0; i < tamanho; i++)
        printf("%d ", dados[(head + i) % MAX]);
    printf("\n");
}

/* ----- main ----- */

int main() {
    enqueue(10);
    enqueue(20);
    enqueue(30);
    imprimir();                             /* 10 20 30 */

    printf("Dequeue: %d\n", dequeue());     /* 10 — liberou posição 0 */
    printf("Dequeue: %d\n", dequeue());     /* 20 — liberou posição 1 */

    enqueue(40);
    enqueue(50);
    /* tail voltou ao início — reutilizou as posições 0 e 1 */
    imprimir();                             /* 30 40 50 */

    enqueue(60);
    enqueue(70);
    imprimir();                             /* 30 40 50 60 70 — array cheio */

    enqueue(80);                            /* "Fila cheia!" */

    return 0;
}

File: FilaCircular.java
public class FilaCircular {

    static int[] dados = new int[1000];
    static int inicio = 0;
    static int fim = 0;
    static int tamanho = 0;
    static boolean vazia(){
        return tamanho == 0;
    }
    static boolean cheia(){
        return fim == 1000;
    }
    static void enfileirar(int x){
        if(cheia()) return;
        dados[fim] = x;
        fim = (fim + 1) % 1000;
        tamanho++;
    }

    static int desenfileirar(){
        if(vazia()) return -1;
        int valor = dados[inicio];
        inicio = (inicio + 1) % 1000;
        tamanho--;
        return valor;
    }
    static void imprimir(){
        for(int i = 0;i< tamanho;i++){
            System.out.print(dados[inicio + i % 1000] + " ");
        }
    }
    public static void main(String[] args){
        enfileirar(10);
        enfileirar(20);
        enfileirar(30);
        desenfileirar();
        imprimir();
    }
}
